### PR TITLE
Reflow curve interface

### PR DESCRIPTION
**Pull request by KPrasch**
- Makes small adjustments to the curve module initialization order and collection to assist with nucyper #370
- Utilize properties to protect internal values while still exposing them publicly
- Additional documentation
- Increased test accuracy and coverage

### CODE REVIEW

File: umbral/curve.py [hunk 1]
```diff
@@ -1,52 +1,69 @@
...
+    def from_name(cls, name: str):
+        return cls(nid=cls._supported_curves[name])
 
     def __eq__(self, other):
-        return self.curve_nid == other.curve_nid
+        return self.__curve_nid == other.curve_nid
 
     def __repr__(self):
-        return "<OpenSSL Curve w/ NID {}>".format(self.curve_nid)
+        return "<OpenSSL Curve(nid={})>".format(self.__curve_nid)
```
Comment (cygnusv) [inline umbral/curve.py]: Can we get the name of the curve here, instead of/in addition to the NID?

File: umbral/curve.py [hunk 2]
```diff
@@ -1,52 +1,69 @@
...
     as a convenient whitelist to limit the curves used in pyUmbral.
     """
 
-    def __init__(self, curve_nid: int):
+    _supported_curves = {
+        'secp256r1': 415,
+        'secp256k1': 714,
+        'secp384r1': 715,
+    }
+
```
Comment (cygnusv) [inline umbral/curve.py]: Perhaps it's more correct wrt to how __init__ works to define the dict the other way around: NIDs as keys, names as values.

File: umbral/curve.py [hunk 3]
```diff
@@ -1,52 +1,69 @@
...
+        return self.__curve_nid == other.curve_nid
 
     def __repr__(self):
-        return "<OpenSSL Curve w/ NID {}>".format(self.curve_nid)
+        return "<OpenSSL Curve(nid={})>".format(self.__curve_nid)
 
     def get_field_order_size_in_bytes(self) -> int:
         backend = default_backend()
         size_in_bits = openssl._get_ec_group_degree(self.ec_group)
         return (size_in_bits + 7) // 8
```
Comment (cygnusv) [inline umbral/curve.py]: This should be cached.

File: umbral/curve.py (hunk 2 above)
Comment (KPrasch) [reply umbral/curve.py]: :heavy_check_mark:

File: umbral/curve.py (hunk 1 above)
Comment (KPrasch) [reply umbral/curve.py]: :heavy_check_mark:

File: umbral/curve.py (hunk 3 above)
Comment (KPrasch) [reply umbral/curve.py]: :heavy_check_mark:

File: .circleci/config.yml [hunk 4]
```diff
@@ -152,7 +152,7 @@ jobs:
           name: Run Mypy Static Type Checks
           command: |
               pipenv install --dev --skip-lock lxml
-              pipenv run mypy --xslt-html-report ./mypy_report ./umbral --config-file=mypy.ini
+              pipenv run mypy --xslt-html-report ./mypy_report ./umbral --config-file=mypy.ini || true
```
Comment (cygnusv) [inline .circleci/config.yml]: What a relief...

Comment (jMyles) [reply .circleci/config.yml]: :relieved:

File: tests/test_curve.py [hunk 5]
```diff
@@ -1,16 +1,16 @@
 import pytest
-from umbral.curve import Curve, SECP256R1, SECP256K1, SECP384R1, _AVAIL_CURVES
+from umbral.curve import Curve, CURVES, SECP256R1, SECP256K1, SECP384R1
 
 
 def test_curve_whitelist():
     # Test the AVAIL_CURVES dict to have only these three curves:
```
Comment (cygnusv) [inline tests/test_curve.py]: AVAIL_CURVES is outdated

File: umbral/curve.py [hunk 6]
```diff
@@ -1,52 +1,105 @@
...
+        714: 'secp256k1',
+        715: 'secp384r1'
+    }
+
+    def __init__(self, nid: int) -> None:
         """
-        Instantiates an OpenSSL curve with the provided curve_nid and derives
+        Instantiates an OpenSSL curve with the provided __curve_nid and derives
         the proper EC_GROUP struct and order. You can _only_ instantiate curves
         with supported nids (see `Curve.supported_curves`).
```
Comment (cygnusv) [inline umbral/curve.py]: `Curve._supported_curves`

File: umbral/curve.py [hunk 7]
```diff
@@ -1,52 +1,105 @@
...
+        # set only once
+        self.__curve_nid = nid
+        self.__ec_group = openssl._get_ec_group_by_curve_nid(self.__curve_nid)
+        self.__order = openssl._get_ec_order_by_group(self.ec_group)
+        self.__generator = openssl._get_ec_generator_by_group(self.ec_group)
+
+    @classmethod
+    def from_name(cls, name: str) -> 'Curve':
+        """
+        Alternate constructor to generate a curve instance by it's name.
```
Comment (cygnusv) [inline umbral/curve.py]: it's --> its

File: umbral/curve.py [hunk 8]
```diff
@@ -1,52 +1,105 @@
...
+    #
+    # Immutable Curve Data
+    #
+
+    @property
     def get_field_order_size_in_bytes(self) -> int:
         backend = default_backend()
-        size_in_bits = openssl._get_ec_group_degree(self.ec_group)
+        size_in_bits = openssl._get_ec_group_degree(self.__ec_group)
         return (size_in_bits + 7) // 8
```
Comment (cygnusv) [inline umbral/curve.py]: Is this being cached?

File: umbral/curve.py (hunk 7 above)
Comment (KPrasch) [reply umbral/curve.py]: :heavy_check_mark:

File: tests/test_curve.py (hunk 5 above)
Comment (KPrasch) [reply tests/test_curve.py]: :heavy_check_mark:

File: umbral/curve.py (hunk 6 above)
Comment (KPrasch) [reply umbral/curve.py]: :heavy_check_mark:

**Pull request merged**